inference-engine/tests/unit/vpu/middleend_tests/passes_tests/reshape_before_conv.cpp

```diff
@@ -0,0 +1,550 @@
+
+#include "ie_memcpy.h"
+
+namespace vpu {
+
+namespace ie = InferenceEngine;
+
+class ReshapeBeforeConvTests : public GraphTransformerTest,
+    public testing::WithParamInterface<std::tuple<DimValues, DimValues>> {
+protected:
+    void SetUp() override {
+        ASSERT_NO_FATAL_FAILURE(GraphTransformerTest::SetUp());
+        ASSERT_NO_FATAL_FAILURE(InitCompileEnv());
+        ASSERT_NO_FATAL_FAILURE(InitPipeline());
+
+        _model = CreateModel();
+    }
+
+    void Compile() {
+        _pipeline.run(_model);
+    }
+
+    void InitPipeline() {
+        _pipeline = PassSet();
+        _pipeline.addPass(passManager->dumpModel("before-convert-shape-notation"));
+        _pipeline.addPass(passManager->reshapeBeforeConvTiling());
+        _pipeline.addPass(passManager->dumpModel("after-convert-shape-notation"));
+    }
+
+    Data InitInputData(const DimValues& dims) {
+        const auto inDesc = DataDesc(DataType::FP16, DimsOrder::NCHW, dims);
+        return _model->addInputData("Input", inDesc);
+    }
+
+    Data InitOutputData(const DimValues& dims) {
+        const auto outDesc = DataDesc(DataType::FP16, DimsOrder::NCHW, dims);
+        return _model->addOutputData("Output", outDesc);
+    }
+
+    Data InitNewData(const DimValues& dims, const std::string& name) {
+        const auto newDesc = DataDesc(DataType::FP16, DimsOrder::NCHW, dims);
+        return _model->addNewData(name, newDesc);
+    }
+
+    void InitConvStage(const Data& input, const Data& output, int kernel_x = 1, int kernel_y = 1) {
+        auto convLayer = std::make_shared < ie::ConvolutionLayer > (ie::LayerParams {
+            "TestConvolution",
+            "StubConv",
+            ie::Precision::FP16
+        });
+        convLayer->_kernel_x = kernel_x;
+        convLayer->_kernel_y = kernel_y;
+        convLayer->_stride_x = 1;
+        convLayer->_stride_y = 1;
+
+        Stage stage = stageBuilder->addConvolutionStage(_model, "TestConvolution", convLayer, input, output,
+                InitNewData(DimValues{ {Dim::N, output->desc().dim(Dim::C)}, {Dim::C, input->desc().dim(Dim::C)}, {Dim::H, kernel_y}, {Dim::W, kernel_x} },
+                "TestConvolution@weights@conv"),
+                _model->addFakeData(), _model->addFakeData());
+
+        stage->attrs().set<int>("kernelSizeX", kernel_x);
+        stage->attrs().set<int>("kernelSizeY", kernel_y);
+        stage->attrs().set<int>("kernelStrideX", 1);
+        stage->attrs().set<int>("kernelStrideY", 1);
+        stage->attrs().set<bool>("tryHW", true);
+    }
+
+    void Validate() {
+        const auto datas = _model->datas();
+        const auto stages = _model->getStages();
+        ASSERT_EQ(datas.size(), pattern_datas.size());
+        ASSERT_EQ(stages.size(), pattern_stages.size());
+
+        auto dataIter = datas.begin();
+        for (auto patternIter = pattern_datas.begin(); patternIter != pattern_datas.end();
+                ++dataIter,
+                ++patternIter) {
```

Comment:
If you need to traverse 2 containers at the same time of the same size - just use index approach - it will simplify code

Reply:
fixed